File: android/src/main/java/com/margelo/nitro/fs2/MediaStore.kt
package com.margelo.nitro.fs2

import com.margelo.nitro.core.Promise
import com.margelo.nitro.fs2.utils.FsError
import com.margelo.nitro.fs2.utils.JsVisibleError
import androidx.core.net.toUri

class MediaStore(): HybridMediaStoreSpec() {
    private val mediaStoreManager = RNFSMediaStoreManager()

    /**
     * Builds the JS-facing error for [ex] without throwing it. See `Fs2.fsError`.
     */
    private fun fsError(context: String, ex: Exception): Throwable {
        // Already a JS-facing error carrying a formatted message - propagate it unchanged.
        if (ex is JsVisibleError) return ex

        // You can expand this for more specific error types as needed
        return FsError(ex.message ?: "Error in MediaStore operation: $context")
    }

    /** Throws the JS-facing error for [ex]. Every path throws, hence [Nothing]. */
    private fun reject(context: String, ex: Exception): Nothing = throw fsError(context, ex)

    override fun mediaStoreCreateFile(
        fileDescription: FileDescription,
        mediaCollection: MediaCollectionType
    ): Promise<String> {
        return Promise.async {
            try {
                val uri = mediaStoreManager.createMediaFile(fileDescription, mediaCollection)
                return@async uri.toString()
            } catch (e: Exception) {
                throw reject(fileDescription.name, e)
            }
        }
    }

    override fun mediaStoreUpdateFile(
        uri: String,
        fileDescription: FileDescription,
        mediaCollection: MediaCollectionType
    ): Promise<String> {
        return Promise.async {
            try {
                val updated = mediaStoreManager.updateMediaFile(uri.toUri(), fileDescription, mediaCollection)
                if (updated) {
                    return@async uri
                } else {
                    throw FsError("Failed to update file: $uri")
                }
            } catch (e: Exception) {
                throw reject(uri, e)
            }
        }
    }

    override fun mediaStoreWriteToFile(uri: String, sourceFilePath: String): Promise<Unit> {
        return Promise.async {
            try {
                val success = mediaStoreManager.writeToMediaFile(uri.toUri(), sourceFilePath)
                if (success) {
                    return@async
                } else {
                    throw FsError("Failed to write to file: $uri")
                }
            } catch (e: Exception) {
                throw reject(uri, e)
            }
        }
    }

    override fun mediaStoreCopyFromFile(
        sourceFilePath: String,
        fileDescription: FileDescription,
        mediaCollection: MediaCollectionType
    ): Promise<String> {
        return Promise.async {
            try {
                val uri = mediaStoreManager.copyToMediaStore(fileDescription, mediaCollection, sourceFilePath)
                return@async uri.toString()
            } catch (e: Exception) {
                throw reject(sourceFilePath, e)
            }
        }
    }

    override fun mediaStoreQueryFile(searchOptions: MediaStoreSearchOptions): Promise<MediaStoreFile?> {
        return Promise.async {
            try {
                // The spec types this `Promise<MediaStoreFile | undefined>`, so "no match" is a
                // resolved `undefined`, not a rejection.
                return@async mediaStoreManager.query(searchOptions)
            } catch (e: Exception) {
                throw reject(searchOptions.fileName ?: "query", e)
            }
        }
    }

    override fun mediaStoreDeleteFile(uri: String): Promise<Boolean> {
        return Promise.async {
            try {
                val deleted = mediaStoreManager.delete(uri.toUri())
                return@async deleted
            } catch (e: Exception) {
                throw reject(uri, e)
            }
        }
    }
}

File: src/_mediastore.ts
import { NitroModules } from 'react-native-nitro-modules';
import type { MediaStore as NitroMediaStore } from './nitro/MediaStore.nitro';

/**
 * Helpers
 */
import { normalizeFilePath } from './utils';

import type {
  MediaCollectionType,
  FileDescription,
  MediaStoreFile,
  MediaStoreSearchOptions,
} from './nitro/MediaStore.nitro';

/**
 * Nitro Hybrid Objects
 */
const RNFSMediaStoreManager =
  NitroModules.createHybridObject<NitroMediaStore>('MediaStore');

/**
 * MediaStore
 */
export default {
  createMediaFile(
    fileDescription: FileDescription,
    mediatype: MediaCollectionType
  ): Promise<string> {
    if (!fileDescription.parentFolder) fileDescription.parentFolder = '';
    return RNFSMediaStoreManager.mediaStoreCreateFile(
      fileDescription,
      mediatype
    );
  },

  updateMediaFile(
    uri: string,
    fileDescription: FileDescription,
    mediatype: MediaCollectionType
  ): Promise<string> {
    return RNFSMediaStoreManager.mediaStoreUpdateFile(
      uri,
      fileDescription,
      mediatype
    );
  },

  writeToMediaFile(uri: string, path: string): Promise<void> {
    return RNFSMediaStoreManager.mediaStoreWriteToFile(
      uri,
      normalizeFilePath(path)
    );
  },

  copyToMediaStore(
    fileDescription: FileDescription,
    mediatype: MediaCollectionType,
    path: string
  ): Promise<string> {
    return RNFSMediaStoreManager.mediaStoreCopyFromFile(
      normalizeFilePath(path),
      fileDescription,
      mediatype
    );
  },

  queryMediaStore(
    searchOptions: MediaStoreSearchOptions
  ): Promise<MediaStoreFile | undefined> {
    return RNFSMediaStoreManager.mediaStoreQueryFile(searchOptions);
  },

  deleteFromMediaStore(uri: string): Promise<boolean> {
    return RNFSMediaStoreManager.mediaStoreDeleteFile(uri);
  },

  MEDIA_AUDIO: 'Audio' as MediaCollectionType,
  MEDIA_IMAGE: 'Image' as MediaCollectionType,
  MEDIA_VIDEO: 'Video' as MediaCollectionType,
  MEDIA_DOWNLOAD: 'Download' as MediaCollectionType,
};
